crates/accelerate/src/sparse_observable.rs

```diff
@@ -0,0 +1,1677 @@
+    ///
+    ///     Construct a simple operator::
+    ///
+    ///         >>> SparseObservable.from_sparse_list(
+    ///         ...     [("ZX", (1, 4), 1.0), ("YY", (0, 3), 2j)],
+    ///         ...     num_qubits=5,
+    ///         ... )
+    ///         <SparseObservable with 2 terms on 5 qubits: (1+0j)(X_4 Z_1) + (0+2j)(Y_3 Y_0)>
+    ///
+    ///     Construct the identity observable (though really, just use :meth:`identity`)::
+    ///
+    ///         >>> SparseObservable.from_sparse_list([("", (), 1.0)], num_qubits=100)
+    ///         <SparseObservable with 1 term on 100 qubits: (1+0j)()>
+    ///
+    ///     This method can replicate the behavior of :meth:`from_list`, if the qubit-arguments
+    ///     field of the triple is set to decreasing integers::
+    ///
+    ///         >>> labels = ["XY+Z", "rl01", "-lXZ"]
+    ///         >>> coeffs = [1.5j, 2.0, -0.5]
+    ///         >>> from_list = SparseObservable.from_list(list(zip(labels, coeffs)))
+    ///         >>> from_sparse_list = SparseObservable.from_sparse_list([
+    ///         ...     (label, (3, 2, 1, 0), coeff)
+    ///         ...     for label, coeff in zip(labels, coeffs)
+    ///         ... ])
+    ///         >>> assert from_list == from_sparse_list
+    #[staticmethod]
+    #[pyo3(name = "from_sparse_list", signature = (iter, /, num_qubits))]
+    fn py_from_sparse_list(
+        iter: Vec<(String, Vec<u32>, Complex64)>,
+        num_qubits: u32,
+    ) -> Result<Self, LabelError> {
+        let coeffs = iter.iter().map(|(_, _, coeff)| *coeff).collect();
+        let mut boundaries = Vec::with_capacity(iter.len() + 1);
+        boundaries.push(0);
+        let mut indices = Vec::new();
+        let mut bit_terms = Vec::new();
+        let mut sorted = btree_map::BTreeMap::new();
+        for (label, qubits, _) in iter {
+            sorted.clear();
+            let label: &[u8] = label.as_ref();
+            if label.len() != qubits.len() {
+                return Err(LabelError::WrongLengthIndices {
+                    label: label.len(),
+                    indices: indices.len(),
+                });
+            }
+            for (letter, index) in label.iter().zip(qubits) {
+                if index >= num_qubits {
+                    return Err(LabelError::BadIndex { index, num_qubits });
+                }
+                let btree_map::Entry::Vacant(entry) = sorted.entry(index) else {
+                    return Err(LabelError::DuplicateIndex { index });
+                };
+                entry.insert(
+                    BitTerm::try_from_u8(*letter).map_err(|_| LabelError::OutsideAlphabet)?,
+                );
+            }
+            for (index, term) in sorted.iter() {
+                let Some(term) = term else {
+                    continue;
+                };
+                indices.push(*index);
+                bit_terms.push(*term);
+            }
+            boundaries.push(bit_terms.len());
+        }
+        Ok(Self {
```

Comment:
Does this call `SparseObservable::new` or does it directly set the values? If it calls `new`, should we call `new_unchecked` instead since we know the data is valid?

Reply:
This constructs the struct directly.  It's effectively the same as `new_unchecked` in this case, but `new_unchecked` was me being slightly forwards-looking to a time when `SparseObservable` is accessed and created from within Rust.  Since the fields of the struct aren't public (deliberately), anywhere outside this file wouldn't be allowed to use the `SparseObservable { .. }` constructor form, but I still wanted the zero-copy, zero-validation form to be available if the user promised to be safe about it.